### DISCUSSION TITLE
Support OpenShift Routes as an alternative to Kubernetes Ingresses

### DISCUSSION THREAD

**Discussion by sergkoudi** [InseeFrLab/legacy-onyxia-entrypoint#65] [Ideas]
In OpenShift clusters, [Routes](https://docs.openshift.com/container-platform/4.11/networking/routes/route-configuration.html) are [preferable](https://cloud.redhat.com/blog/kubernetes-ingress-vs-openshift-route) to Kubernetes Ingresses. When running Onyxia on OpenShift clusters, it requires Routes configuration as an alternative to Ingresses.

I am suggesting providing Routes configuration for all the Onyxia services' charts of the services that expose the URL to the outside of the cluster.

## Implementation steps

1. In the [region configuration](https://github.com/InseeFrLab/onyxia-api/blob/master/docs/region-configuration.md#services-properties), add support of OpenShift in the property `regions.services.type` with the new value `OPENSHIFT`.
2. In the `values.schema.json` files, move `ingress` configuration to the inside of `networking` (see [example config](https://github.com/InseeFrLab/helm-charts-interactive-services/blob/main/charts/vscode-python/values.schema.json#L518-L593)).
3. Add a hidden `route` configuration similar to `ingress`.
4. Add choosing between `ingress` or `routes` for Onyxia users, defaulting to `ingress`. It can be a radio button on the interface. It must appear when `regions.services.type` is `OPENSHIFT` (this condition is useful because Kubernetes clusters can't use OpenShift Routes, according to my quick research). The interface example is here:
![External access](https://user-images.githubusercontent.com/23199590/207908024-d0cb89d1-e132-4633-8190-350b3a2c692b.jpeg)
5. Make the corresponding changes to chart values and templates providing templates for routes.

## Questions

- Can 2 different regions be used by a single user? If yes, how the Onyxa admin choses between them when launching a service? The question is related to the conditional networking described above.

**Comment by fcomte** [comment]
do you want that a user can choose between ingress and route ? 
It seems that the choice should never appear to the end user... 

There is some code to do in the API to analyse the route manifest and give all urls to the end users.

**Reply by sergkoudi** [reply]
> do you want that a user can choose between ingress and route ?

Yes, because OpenShift can also use Kubernetes Ingresses, moreover at the same time.

**Reply by fcomte** [reply]
For regions :

It's not possible to select the regions in the web ui at the moment. But we will add that.
We want that a user can select a region in the UI 

And maybe after providing some tools to help users to manage data across that two regions.

**Reply by sergkoudi** [reply]
Indeed, we can fully delegate this to kubernetes cluster admins. Probably somewhere  in the regions configuration.

**Reply by fcomte** [reply]
Yes i guess choosing between route or ingress is more an admin business. 
Users should not see that in most case.

**Reply by sergkoudi** [reply]
In that case, I suggest adding the property `regions.services.expose.type` with possible values `ingress` and `route`.

Should we protect somehow setting this value to `route` when `regions.services.type` is `KUBERNETES`?

**Reply by fcomte** [reply]
`regions.services.type` was used to separate `KUBERNETES` and `MARATHON (mesos)` containers. 

I think we can only use `regions.services.expose.type`

**Comment by fcomte** [comment]
You talk about admins where we talk about users. 
Do you mean that a onyxia user is an admin ?

**Reply by sergkoudi** [reply]
corrected to users
